### PR TITLE
Upgrades pgx dependency to v3

### LINKED ISSUE #18: Support for pgx 3.0
Is it planned?

### PR DESCRIPTION
**Pull request by aravindc26**
Most changes are related to [pgtype](https://godoc.org/github.com/jackc/pgx/pgtype) .

Fixes #18.

### CODE REVIEW

File: work_test.go [hunk 1]
```diff
@@ -238,21 +239,19 @@ func TestLockJobAdvisoryRace(t *testing.T) {
...
 		}
 		return backendID
 	}
-	waitUntilBackendIsWaiting := func (backendID int32, name string) {
+	waitUntilBackendIsWaiting := func(backendID int32, name string) {
 		conn := newConn()
 		i := 0
 		for {
 			var waiting bool
-			err := conn.QueryRow(`SELECT pg_stat_get_backend_waiting($1)`, backendID).Scan(&waiting)
```
Comment (aravindc26) [inline work_test.go]: postgres 9.6 doesn't seem to support `pg_stat_get_backend_waiting`

Comment (cbandy) [inline work_test.go]: We should get some CI going that tests different PostgreSQL versions...

File: work_test.go [hunk 2]
```diff
@@ -238,21 +239,19 @@ func TestLockJobAdvisoryRace(t *testing.T) {
 	getBackendID := func(conn *pgx.Conn) int32 {
 		var backendID int32
 		err := conn.QueryRow(`
-			SELECT backendid
-			FROM pg_stat_get_backend_idset() psgb(backendid)
-			WHERE pg_stat_get_backend_pid(psgb.backendid) = pg_backend_pid()
+			SELECT pg_backend_pid()
 		`).Scan(&backendID)
```
Comment (cbandy) [inline work_test.go]: The value here is changing from the "backend ID" used by the [per-backend stats functions](https://www.postgresql.org/docs/9.6/static/monitoring-stats.html#MONITORING-STATS-BACKEND-FUNCS-TABLE) to the "process ID." I think the names should change accordingly. Perhaps `backendPID` here and elsewhere?

**Comment by pims** [comment]
In the meantime I submitted a PR to [pin dependencies to v2](https://github.com/bgentry/que-go/pull/17)

File: que.go [hunk 3]
```diff
@@ -155,38 +156,53 @@ func execEnqueue(j *Job, q queryable) error {
...
 
-	queue := pgx.NullString{
+	var status pgtype.Status
+	if j.Queue != "" {
+		status = pgtype.Present
+	} else {
+		status = pgtype.Null
+	}
+
+	queue := &pgtype.Text{
```
Comment (aravindc26) [inline que.go]: @cbandy its required that pgtype values be pointers, please refer [here](https://github.com/jackc/pgx/issues/292).

File: work_test.go [hunk 4]
```diff
@@ -238,21 +239,19 @@ func TestLockJobAdvisoryRace(t *testing.T) {
...
 		return backendID
 	}
-	waitUntilBackendIsWaiting := func (backendID int32, name string) {
+	waitUntilBackendIsWaiting := func(backendID int32, name string) {
 		conn := newConn()
 		i := 0
 		for {
 			var waiting bool
-			err := conn.QueryRow(`SELECT pg_stat_get_backend_waiting($1)`, backendID).Scan(&waiting)
+			err := conn.QueryRow(`SELECT wait_event is not null from pg_stat_activity where pid=$1`, backendID).Scan(&waiting)
```
Comment (aravindc26) [inline work_test.go]: `pg_stat_get_backend_waiting` could return true if the process is waiting for a lock, false if it is not waiting or null if the process id is invalid. The test would panic when null is returned.

In the changed query `ErrNoRows` is possible if the backendPID does not exist, in that case is it not appropriate to panic ?

Comment (cbandy) [inline work_test.go]: I didn't realize the function could return null. Panic is okay here.

File: work_test.go [hunk 5]
```diff
@@ -236,23 +237,21 @@ func TestLockJobAdvisoryRace(t *testing.T) {
 		return conn
 	}
 	getBackendID := func(conn *pgx.Conn) int32 {
```
Comment (bgentry) [inline work_test.go]: can you also rename `getBackendID` to `getBackendPID` for consistency with the variable rename (which I agree was the right choice)?

File: work_test.go (hunk 1 above)
Comment (bgentry) [inline work_test.go]: yeah, looks like it was removed. Compare the [9.5 docs](https://www.postgresql.org/docs/9.5/static/monitoring-stats.html#MONITORING-STATS-BACKEND-FUNCS-TABLE) to the [9.6 docs](https://www.postgresql.org/docs/9.6/static/monitoring-stats.html#MONITORING-STATS-BACKEND-FUNCS-TABLE).

Can you briefly explain how these are equivalent?

File: work_test.go [hunk 6]
```diff
@@ -608,7 +607,7 @@ func TestJobError(t *testing.T) {
 	}
 	defer j2.Done()
 
-	if !j2.LastError.Valid || j2.LastError.String != msg {
+	if (j2.LastError.Status == pgtype.Null) || j2.LastError.String != msg {
```
Comment (bgentry) [inline work_test.go]: parentheses are not necessary here

File: work_test.go (hunk 1 above)
Comment (aravindc26) [inline work_test.go]: @bgentry In 9.5 `pg_stat_get_backend_waiting` was a helper function that fetches value from `pg_stat_activity.waiting` column. In 9.6 the column was removed and was replaced by wait_event, which is more explanatory than a bool column; see (https://www.postgresql.org/docs/9.6/static/monitoring-stats.html#WAIT-EVENT-TABLE) for all possible values of `wait_event`. `wait_event` is null if the process does not hold any lock.

**Comment by aravindc26** [comment]
@bgentry I have made the changes.

**Comment by bgentry** [comment]
@aravindc26 I'm seeing three more minor changes that haven't yet been addressed. Sorry for the delay, I'm able to move fast on this this week.

**Comment by aravindc26** [comment]
@bgentry can you please point it out ? I believe I have resolved everything.

**Comment by bgentry** [comment]
@aravindc26 I think GitHub's UI was playing tricks on me. Looks good now, thanks! released as v1.0.0

**Pull request merged**